11 avoid specific ips

sancho20021 opened the pull request:
# Description
Add new command that specifies what IPs or range of IPs to avoid when generating IPs for peers.
## Related issues
Fixes #11 
## Solution explanation
For every network of the interface compute the difference between the network and the IP range that we want to ignore. That is done by function 
```
ban_ip_subnet(ns: &Vec<IpNetwork>, banned: &IpNetwork) -> Vec<IpNetwork>
```
Then, when allocating an IP for a user with specified index `n`, instead of taking `nth` IP inside the network, iterate over non-ignored IP ranges until we skip `n` IPs.

sancho20021 committed:
[#11] Allow ignoring some IP ranges when generating IPs for peers

Problem: No functionality to avoid specific IPs when generating
IPs for peers.

Solution: Add `ignore` command (only IPv4 for now).

sancho20021 committed:
[#11] Implement `ignore` for IPv6

Problem: `ignore` command was implemented only for IPv4.

Solution: implement for IPv6.

cab404 commented on src/configs/mod.rs:
```diff
@@ -451,6 +455,33 @@ impl WireguardNetworkInfo {
     }
+
+    pub fn get_free_net_address(&self, net: IpNetwork, num: u128) -> IpAddr {
+        let mut cum = 0u128;
```
sorry, what?

sancho20021 replied:
cumulative sum. Ok I'll rename it somehow

cab404 commented on src/configs/mod.rs:
```diff
@@ -483,3 +514,71 @@ pub trait ConfigType {
     fn write_config(net: WireguardConfiguration, options: Self::ExportConfig) -> String;
 }
+
+pub fn ban_ip_subnet(ns: &Vec<IpNetwork>, banned: &IpNetwork) -> Vec<IpNetwork> {
```
this can probably be replaced by `map` call

sancho20021 replied:
Do you mean map and then concat? Because `exclude_subnet` has type `IpNetwork -> Vec<IpNetwork>`.

sancho20021 replied:
`ns.iter().flat_map(|n| exclude_subnet(*n, banned)).collect()` ?

cab404 replied:
ns.iter().flat_map(|n| exclude_subnet(*n, banned)).collect() ?

yeah, this

cab404 commented on src/configs/mod.rs:
```diff
@@ -451,6 +455,33 @@ impl WireguardNetworkInfo {
+            cum += size(&subnet);
+        }
+        //TODO: Consider showing message that no more IPs in net left
+        panic!()
```
let’s at least add a message to it)

cab404 commented:
I kinda don’t understand, why do we need considered networks?
It looks like id "jumping" between considered networks mechanic makes it less understandable by mere humans — and just skipping ids according to the ignore list might be better?

it would also free you from implementing subnet subtraction op (however it’s cool you did them — we can now replace the list of  "IPv6" and "IPv4" public networks with a list of private ones (which is easier to obtain), and you also can PR those to [ipnetwork library](https://github.com/achanda/ipnetwork/))

sancho20021 committed:
[#11] Rewrite implementation of IP ranges ignoring

Problem: Current implementation is quite hard to understand.

Solution: Store reserved peer IPs in hashmap.

sancho20021 commented:
> I kinda don’t understand, why do we need considered networks? It looks like id "jumping" between considered networks mechanic makes it less understandable by mere humans — and just skipping ids according to the ignore list might be better?
> 
> it would also free you from implementing subnet subtraction op (however it’s cool you did them — we can now replace the list of "IPv6" and "IPv4" public networks with a list of private ones (which is easier to obtain), and you also can PR those to [ipnetwork library](https://github.com/achanda/ipnetwork/))

Ok, agree. I'v rewritten it other way. Didn't support IPv6 yet, If you approve this version I'll add implementation for IPv6.

cab404 commented:
Oooof, I've just noticed the problem of supporting several networks and allocating IPs in each of them — I guess we’ll be dropping ID→IP mapping (sadly)

However, we probably want to move IPs to peer info, instead of saving it in the network config — can you do that?

sancho20021 commented:
> Oooof, I've just noticed the problem of supporting several networks and allocating IPs in each of them — I guess we’ll be dropping ID→IP mapping (sadly)
> 
> However, we probably want to move IPs to peer info, instead of saving it in the network config — can you do that?

Yes, sure. Btw this mapping is already dropped because of ignored IP ranges. Or, it is deterministic but too complex. 
So, for new peers, we should just allocate some free unreserved IP and store it in `PeerInfo`?

cab404 commented:
> So, for new peers, we should just allocate some free unreserved IP and store it in `PeerInfo`?

yeah

sancho20021 commented:
> > So, for new peers, we should just allocate some free unreserved IP and store it in `PeerInfo`?
> 
> yeah

But we have a list of networks. Isn't it strange that `PeerInfo` will contain information about the networks (If we store `HashMap<IpNetwork, IpAddr>` in `PeerInfo`)?

cab404 commented:
> But we have a list of networks. Isn't it strange that `PeerInfo` will contain information about the networks (If we store `HashMap<IpNetwork, IpAddr>` in `PeerInfo`)?

Not really. That was network-wide info, peer info is here to store 1-1 peer info, and IP list is such data

sancho20021 commented:
> > But we have a list of networks. Isn't it strange that `PeerInfo` will contain information about the networks (If we store `HashMap<IpNetwork, IpAddr>` in `PeerInfo`)?
> 
> Not really. That was network-wide info, peer info is here to store 1-1 peer info, and IP list is such data

I'm not sure I understand what do you mean( 
1. For each peer and for each network we need to know peer's IP in this network, right? Then, what does __1-1__ mean?
2. What object we need to store inside `PeerInfo`? One IP (then from which network), List of IPs, Map from Network to IP?

cab404 commented:
> > > But we have a list of networks. Isn't it strange that `PeerInfo` will contain information about the networks (If we store `HashMap<IpNetwork, IpAddr>` in `PeerInfo`)?
> > 
> > 
> > Not really. That was network-wide info, peer info is here to store 1-1 peer info, and IP list is such data
> 
> I'm not sure I understand what do you mean(
> 
>     1. For each peer and for each network we need to know peer's IP in this network, right? Then, what does **1-1** mean?
> 
>     2. What object we need to store inside `PeerInfo`? One IP (then from which network), List of IPs, Map from Network to IP?
1) we don’t really need to know network for IPs
2) just a list of IPs

sancho20021 committed:
[#11] Move peer IPs from WireguardNetworkInfo to PeerInfo

Problem: Ignored ranges and multiple IP networks introduced
difficulties in Peer ID -> IP mapping.

Solution: Allocate new peer IPs without using id and store them
in PeerInfo.

sancho20021 commented:
> 1. we don’t really need to know network for IPs
> 2. just a list of IPs

Done (only for IPv4)

cab404 commented on Cargo.toml:
```diff
@@ -18,7 +18,7 @@ log = "0.4"
 rand = "0.8.4"
 base64 = "0.13.0"
-bimap = { version = "0.6.2", features = ["serde"] }
+apply = "*"
```
can you pin the version, please?)

sancho20021 replied:
Yes

cab404 reviewed (commented):
everything looks really nice, thank you!
now we just need ipv6, and it’ll be pretty much it

you can also add some tests, if you want to

sancho20021 committed:
[#11] Support IPv6

sancho20021 commented:
> you can also add some tests, if you want to

I suppose testing requires a sequence of `wg-bond` executions with specified arguments, and then checking the properties of the _configuration json_. What can be the preferable way of doing that?

cab404 commented:
> I suppose testing requires a sequence of `wg-bond` executions with specified arguments

Nope, I mean unit tests — create some interesting exclusion rules, check that those are processed fine. For instance, create a config with no networks, and try to allocate something; Create a config with one network and exclude every address except one — and then try allocating address; and so on)

sancho20021 committed:
[#11] Add tests and refactor code

Problem: no tests for ignoring ip ranges.

Solution: add tests. Find bugs and fix them.

cab404 commented:
@sancho20021 tests fail for some reason

cab404 commented:
`wg-bond> thread 'tests::test_ignore_ipv4' panicked at 'called `Result::unwrap()` on an `Err` value: "Failed to run 'wg genkey': No such file or directory (os error 2)"', src/main.rs:604:33`

Hm, I thought I've merged bundled key derivation D:

cab404 commented:
i've rebased changes on top of current master and updated clap and stuff

cab404 commented:
Mind if I also remove all the unused functions?

cab404 commented:
Also, if you create two ignored subnets and then cover both of them with a new rule, only one warning is shown
```
cargo run --  ignore 1.1.1.0/24
cargo run --  ignore 1.1.2.0/24
cargo run --  ignore 1.1.0.0/16
```

cab404 commented:
"unignore" command would be a nice to have)

sancho20021 commented:
> Mind if I also remove all the unused functions?

I left the function for IPNetwork difference, now they are unused but if we do `unignore` command, it'll be needed I think.

sancho20021 commented:
> "unignore" command would be a nice to have)

Ok.
Btw, currently peer receives his IPs right after creating. So, If some IP range is marked as ignored, IPs of existing peers don't change. How do we deal with this situation? Some variants:
- Leave as is, let the user first add all ignored ranges, and then add peers
- Update all IPs each time when new IP range becomes ignored / unignored.
- Have separate command for peer's IPs updating

sancho20021 committed:
fixup! [#11] Add tests and refactor code

cab404 commented:
>  How do we deal with this situation?

Looks to me, that the most reasonable will be to warn that there are IPs in that range, and cancel operation. 
Later, we'll probably add proper change ips/flags/whatever operations.

sancho20021 committed:
[#11] Add ignore command aborting

Problem: IP range is allowed to be ignored even if it contains
assigned IPs.

Solution: Abort ignore operation if necessary.

sancho20021 committed:
[#11] Implement `unignore` command

Problem: Sometimes it's more convenient to ignore everything
except specific IP range.

Solution: Implement `unignore` command by using already implemented
networks subtraction.

sancho20021 commented:
`Unignore` command done.

cab404 commented on src/configs/mod.rs:
```diff
@@ -483,3 +526,81 @@ pub trait ConfigType {
+    fn subtract(&self, other: &Self) -> HashSet<Self>;
+    fn subnets(&self) -> (Self, Self);
+
+    fn subtract_all(minuend: &HashSet<Self>, subtrahend: &Self) -> HashSet<Self> {
```
TIL
